protocols/kad/src/query/peers/disjoint_closest.rs

```diff
@@ -0,0 +1,427 @@
+    /// in order to finish successfully. Defaults to `K_VALUE`.
+    pub num_results: usize,
+
+    /// The timeout for a single peer.
+    ///
+    /// If a successful result is not reported for a peer within this timeout
+    /// window, the iterator considers the peer unresponsive and will not wait for
+    /// the peer when evaluating the termination conditions, until and unless a
+    /// result is delivered. Defaults to `10` seconds.
+    pub peer_timeout: Duration,
+}
+
+impl Default for DisjointClosestPeersIterConfig {
+    fn default() -> Self {
+        DisjointClosestPeersIterConfig {
+            parallelism: ALPHA_VALUE.get(),
+            disjoint_paths: 3,
+            num_results: K_VALUE.get(),
+            peer_timeout: Duration::from_secs(10),
+        }
+    }
+}
+
+/// Wraps around a set of `ClosestPeersIter`, enforcing the amount of configured disjoint paths
+/// according to the S/Kademlia paper.
+pub struct DisjointClosestPeersIter {
+    iters: Vec<ClosestPeersIter>,
+    /// Mapping of yielded peers to iterator that yielded them.
+    ///
+    /// More specifically index into the `DisjointClosestPeersIter::iters` vector. On the one hand
+    /// this is used to link responses from remote peers back to the corresponding iterator, on the
+    /// other hand it is used to track which peers have been contacted in the past.
+    yielded_peers: HashMap<PeerId, usize>,
+}
+
+impl DisjointClosestPeersIter {
+    /// Creates a new iterator with a default configuration.
+    pub fn new<I>(target: KeyBytes, known_closest_peers: I) -> Self
+    where
+        I: IntoIterator<Item = Key<PeerId>>,
+    {
+        Self::with_config(
+            DisjointClosestPeersIterConfig::default(),
+            target,
+            known_closest_peers,
+        )
+    }
+
+    /// Creates a new iterator with the given configuration.
+    pub fn with_config<I, T>(
+        config: DisjointClosestPeersIterConfig,
+        target: T,
+        known_closest_peers: I,
+    ) -> Self
+    where
+        I: IntoIterator<Item = Key<PeerId>>,
+        T: Into<KeyBytes> + Clone,
+    {
+        let iters = split_inputs_per_disjoint_path(&config, known_closest_peers)
+            .into_iter()
+            .map(|(config, peers)| ClosestPeersIter::with_config(config, target.clone(), peers))
+            .collect();
+
+        DisjointClosestPeersIter {
+            iters,
+            yielded_peers: HashMap::new(),
+        }
+    }
+
+    pub fn on_failure(&mut self, peer: &PeerId) {
+        self.iters[self.yielded_peers[peer]].on_failure(peer);
+    }
+
+    pub fn on_success<I>(&mut self, peer: &PeerId, closer_peers: I)
+    where
+        I: IntoIterator<Item = PeerId>,
+    {
+        self.iters[self.yielded_peers[peer]].on_success(peer, closer_peers);
+    }
+
+    pub fn is_waiting(&self, peer: &PeerId) -> bool {
+        self.iters[self.yielded_peers[peer]].is_waiting(peer)
+    }
+
+    pub fn next(&mut self, now: Instant) -> PeersIterState {
```

Comment:
This looks like it's providing an API for the end user to transparently blockingly iterate through the results as they come in on the network, but also implementing the protocol within this type of wrapper.

On the surface it might look like a reasonable thing to do, but it gives you very little control into the low-level details of how the protocol behaves when executed. For example there is no way to make decisions based on whether you received the next hop in path A or path B first. As you noted "[arbitrary ordering] might be unfair". Blocking paradigms such as this make it easier to implement control flows that "look like" multiple unrelated threads, but for more predictable performance we'd like to have control flows that are more interconnected, that don't "look like" multiple unrelated threads.

Implementing the protocol as a non-blocking state machine gives you more control into the details of the process and allows for easier reasoning about what resources are being used at any given time. I would suggest implementing the protocol like this, and then wrapping a blocking iterator around this for the end user, rather than doing the protocol lookups in the iterator-wrapper itself. For example the implementation can just shove the results into a blocking queue, and the end user can consume the queue. Then all the blocking would be localised into the queue implementation, and the performance / resource-usage behaviour of your non-blocking code would be easier to reason about.

This is how my [python example](https://github.com/w3f/research/blob/master/docs/polkadot/networking/skad.py#L8) is implemented, everything is non-blocking there.

Reply:
Another advantage of the non-blocking statemachine approach is it's easier to introspect the memory usage; harder to do that when your protocol state is implicitly the local variables of a bunch of green threads fed into a global executor, mixed together with other protocols.

Reply:
It could be that I'm misunderstanding how the iteration works - e.g. perhaps it is repeatedly polling the child iterators. However this just opens up extra questions - Rust futures are supposed to be woken up when the status changes, but I see no-where in the current PR where a wake-up of the child iterators would forward the wake-up onto the parent iterator, which would be needed under the polling paradigm.

Reply:
I am not entirely sure I understand all of your comments. Let me try to explain some parts that might cause some confusion.

The `libp2p-kad` crate is a synchronous state machine. It takes as input events from the network (e.g. new node connected, response from node x, ...) and produces as output requests to be send out on the network (e.g. send `FIND_NODE` request to node x).

I am guessing by _blocking_ you mean e.g. blocking the current thread until a response was received on a TCP socket. Under this assumption Kademlia has no way of blocking, given that it does not control any network resources (e.g. a Tcp socket), but can only return action requests (e.g. please dial this node, please send this packet to this node, ...).

In regards to any of the `XPeersIterator`: All of their methods are synchronous and single-threaded. While they are blocking in terms of not being asynchronous, they are non-blocking in terms of parking the OS-thread (e.g. blocking on a Tcp connection). You can see each method call as being instant (only bound by the CPU computation time they need).

Does this make sense to you @infinity0?